ethereum/contracts/relayer/wormholeRelayer/WormholeRelayerGovernance.sol

```diff
@@ -0,0 +1,231 @@
+error InvalidFork();
+error ContractUpgradeFailed(bytes failure);
+
+abstract contract WormholeRelayerGovernance is WormholeRelayerBase, ERC1967Upgrade {
+    //This constant should actually be defined in IWormhole. Alas, it isn't.
+    uint16 private constant WORMHOLE_CHAINID_UNSET = 0;
+
+    /**
+     * Governance VMs are encoded in a packed fashion using the general wormhole scheme:
+     *   GovernancePacket = <Common Header|Action Parameters>
+     *
+     * For a more detailed explanation see here:
+     *   - https://docs.wormhole.com/wormhole/governance
+     *   - https://github.com/wormhole-foundation/wormhole/blob/main/whitepapers/0002_governance_messaging.md
+     */
+
+    //Right shifted ascii encoding of "WormholeRelayer"
+    bytes32 private constant module =
+        0x0000000000000000000000000000000000576f726d686f6c6552656c61796572;
+
+    /**
+     * The choice of action enumeration and parameters follows the scheme of the core bridge:
+     *   - https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/BridgeGovernance.sol#L115
+     */
+
+    /**
+     * Registers a wormhole relayer contract that was deployed on another chain with the WormholeRelayer on
+     *   this chain. The equivalent to the core bridge's registerChain action.
+     *
+     * Action Parameters:
+     *   - uint16 foreignChainId 
+     *   - bytes32 foreignContractAddress 
+     */
+    uint8 private constant GOVERNANCE_ACTION_REGISTER_WORMHOLE_RELAYER_CONTRACT = 1;
+
+    /**
+     * Upgrades the WormholeRelayer contract to a new implementation. The equivalent to the core bridge's
+     *   upgrade action.
+     *
+     * Action Parameters:
+     *   - bytes32 newImplementation
+     */
+    uint8 private constant GOVERNANCE_ACTION_CONTRACT_UPGRADE = 2;
+
+    /**
+     * Sets the default relay provider for the WormholeRelayer. Has no equivalent in the core bridge.
+     *
+     * Action Parameters:
+     *   - bytes32 newProvider 
+     */
+    uint8 private constant GOVERNANCE_ACTION_UPDATE_DEFAULT_PROVIDER = 3;
+
+    //By checking that only the contract can call itself, we can enforce that the migration code is
+    //  executed upon program upgrade and that it can't be called externally by anyone else.
+    function checkAndExecuteUpgradeMigration() external {
+        assert(msg.sender == address(this));
+        executeUpgradeMigration();
+    }
+
+    function executeUpgradeMigration() internal virtual {
+        //override and implement in WormholeRelayer upon contract upgrade (if required)
+    }
+
+    function registerWormholeRelayerContract(bytes memory encodedVm) external {
+        (uint16 foreignChainId, bytes32 foreignAddress) =
+            parseAndCheckRegisterWormholeRelayerContractVm(encodedVm);
+
+        getRegisteredWormholeRelayersState().registeredWormholeRelayers[foreignChainId] =
+            foreignAddress;
+    }
+
+    function submitContractUpgrade(bytes memory encodedVm) external {
```

Comment:
Currently, the Token Bridge and Wormhole contracts emit an `event` when the contract is upgraded. Should we consider emitting the same event here?